File: jianzhi_offer/No.12 矩阵中的路径/solution.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        for i in range(len(board)):
            for j in range(len(board[0])):
                if self.dfs(board, word, i, j, 0): return True
        return False
    def dfs(self,b,w,i,j,k):
        if not 0<=i<len(b) or not 0<=j<len(b[0]) or b[i][j]!=w[k]:
            return False 
        if k==len(w)-1: return True
        tmp,b[i][j]=b[i][j],'/'
        res=self.dfs(b,w,i-1,j,k+1) or self.dfs(b,w,i+1,j,k+1) or self.dfs(b,w,i,j-1,k+1) or self.dfs(b,w,i,j+1,k+1)
        b[i][j]=tmp
        return res

File: jianzhi_offer/No.12 矩阵中的路径/solution.cpp
//思路：深度优先搜索算法，递归每个点四个方向搜索，经过的点标记
class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for(int i=0;i<board.size();i++){
            for (int j=0;j<board[0].size();j++){
                if(dfs(board,i,j,0,word)) return true;
            }
        }
        return false;

    }
    bool dfs(vector<vector<char>>& board,int i,int j,int k,string& word){
        if (i >= board.size() || i < 0 || j >= board[0].size() || j < 0 || board[i][j] != word[k]) return false;
        if (k==word.size()-1) return true;
        char tmp=board[i][j];
        board[i][j]='/';
        bool res=dfs(board,i-1,j,k+1,word)||dfs(board,i+1,j,k+1,word)||dfs(board,i,j-1,k+1,word)||dfs(board,i,j+1,k+1,word);
        board[i][j]=tmp;
        return res;
    }
};